File: background.py
"""Detached per-user tunnel supervisor and bounded, authenticated local JSON IPC."""
from __future__ import annotations

import argparse
from dataclasses import asdict
import json
import os
from pathlib import Path
import secrets
import socket
import socketserver
import subprocess
import sys
import threading
import time

from forwarding import DATA_DIR, Forward, InstanceLock, Store, TunnelManager, validate_host

PROTOCOL = 1
MAX_RESPONSE = 2 * 1024 * 1024


def exchange(directory: Path, command: str, **arguments) -> dict:
    endpoint = json.loads((directory / "endpoint.json").read_text(encoding="utf-8"))
    request = {"token": endpoint["token"], "protocol": PROTOCOL, "command": command, **arguments}
    with socket.create_connection(("127.0.0.1", endpoint["port"]), timeout=2) as connection:
        connection.settimeout(5)
        connection.sendall(json.dumps(request).encode("utf-8") + b"\n")
        with connection.makefile("rb") as response:
            raw = response.readline(MAX_RESPONSE + 1)
        if len(raw) > MAX_RESPONSE or not raw.endswith(b"\n"):
            raise OSError("Invalid response from background manager")
        result = json.loads(raw)
        if not result.get("ok"):
            raise OSError(result.get("error", "Background manager rejected the request"))
        return result


def launch_daemon(directory: Path):
    """Break away from the Terminal job and console, inheriting no open handles."""
    directory.mkdir(parents=True, exist_ok=True)
    # The daemon needs only the standard library. Avoid the Windows venv launcher.
    executable = getattr(sys, "_base_executable", sys.executable)
    command = [executable, str(Path(__file__).resolve()), "--serve", "--data-dir", str(directory.resolve())]
    flags = (subprocess.DETACHED_PROCESS | subprocess.CREATE_NEW_PROCESS_GROUP
             | subprocess.CREATE_BREAKAWAY_FROM_JOB)
    log_path = directory / "background.log"
    if log_path.exists() and log_path.stat().st_size > 256 * 1024:
        log_path.replace(directory / "background.previous.log")
    with log_path.open("ab") as log:
        try:
            process = subprocess.Popen(command, stdin=subprocess.DEVNULL, stdout=log, stderr=log,
                                       close_fds=True, creationflags=flags)
        except PermissionError as error:
            raise OSError("Windows prevented background detachment. Launch this app from a normal Windows Terminal session.") from error
    return process


def ensure_daemon(directory: Path) -> dict:
    try:
        return exchange(directory, "status")
    except (OSError, ValueError, KeyError):
        pass
    # A second launcher cannot create a second daemon; the server holds daemon.lock.
    process = launch_daemon(directory)
    deadline = time.monotonic() + 8
    try:
        while time.monotonic() < deadline:
            try:
                return exchange(directory, "status")
            except (OSError, ValueError, KeyError):
                time.sleep(.1)
        raise OSError(f"Background manager did not start. See {directory / 'background.log'}")
    finally:
        # poll reaps a losing startup process without terminating the detached winner.
        process.poll()


class DaemonClient:
    persistent = True

    def __init__(self, host: str, directory: Path):
        self.host, self.directory = host, Path(directory)
        self.running: dict[str, bool] = {}
        self.states: dict[str, str] = {}
        self.logs: dict[str, str] = {}
        self._apply(ensure_daemon(self.directory))

    def _apply(self, result: dict):
        if result.get("host") != self.host:
            raise OSError("The background manager uses a different SSH host. Stop it before changing hosts.")
        self.running = dict.fromkeys(result["running"], True)
        self.states = result["states"]
        self.logs = result["details"]

    def _call(self, command: str, **arguments):
        self._apply(exchange(self.directory, command, **arguments))

    def poll(self):
        try:
            self._call("status")
        except (OSError, ValueError, KeyError) as error:
            for rule_id in self.running:
                self.states[rule_id] = "ERROR"
                self.logs[rule_id] = "Background manager disconnected. Reopen the app to reconnect."
            self.running = {}
            raise OSError(str(error)) from error

    def start(self, rule: Forward):
        try:
            self._call("start", rule_id=rule.id)
        except (OSError, ValueError, KeyError) as error:
            self.states[rule.id] = "ERROR"
            self.logs[rule.id] = str(error)

    def stop(self, rule_id: str):
        self._call("stop", rule_id=rule_id)

    def close(self):
        """Explicit stop-all action. UI detach intentionally does not call this."""
        self._call("stop_all")

    def status(self, rule_id: str) -> str:
        return self.states.get(rule_id, "OFF")

    def details(self, rule_id: str) -> str:
        return self.logs.get(rule_id, "")


class Supervisor:
    def __init__(self, store: Store, manager: TunnelManager):
        self.store, self.manager = store, manager
        self.token = secrets.token_hex(32)
        self.lock = threading.RLock()
        self.stopping = threading.Event()

    def snapshot(self):
        return {"ok": True, "protocol": PROTOCOL, "pid": os.getpid(), "host": self.manager.host,
                "running": list(self.manager.running), "states": dict(self.manager.states),
                "details": {key: self.manager.details(key)[-6000:] for key in self.manager.states}}

    def dispatch(self, request: dict) -> dict:
        token = request.get("token")
        if not isinstance(token, str) or not secrets.compare_digest(token, self.token):
            return {"ok": False, "error": "Unauthorized"}
        if request.get("protocol") != PROTOCOL:
            return {"ok": False, "error": "Incompatible background protocol; restart the background manager"}
        with self.lock:
            action = request.get("command")
            if action == "start":
                self.store.load()
                if self.store.host != self.manager.host:
                    raise ValueError("SSH host changed. Stop the background manager before switching hosts.")
                rule = next((r for r in self.store.forwards if r.id == request.get("rule_id")), None)
                if rule is None:
                    raise ValueError("Save the forward before starting it.")
                self.manager.start(rule)
            elif action == "stop":
                rule_id = request.get("rule_id")
                if not isinstance(rule_id, str):
                    raise ValueError("Invalid forward ID")
                self.manager.stop(rule_id)
            elif action in ("stop_all", "shutdown"):
                self.manager.close()
                if action == "shutdown":
                    self.stopping.set()
            elif action != "status":
                raise ValueError("Unknown command")
            self.manager.poll()
            return self.snapshot()


class LocalServer(socketserver.ThreadingTCPServer):
    daemon_threads = True
    allow_reuse_address = False


def serve(directory: Path):
    directory = directory.resolve()
    lock = InstanceLock(directory, "daemon.lock")
    manager = server = None
    server_started = False
    endpoint_path = directory / "endpoint.json"
    try:
        store = Store(directory)
        store.load()
        validate_host(store.host)
        manager = TunnelManager(store.host, directory)
        supervisor = Supervisor(store, manager)

        class Handler(socketserver.StreamRequestHandler):
            def handle(self):
                self.connection.settimeout(2)
                try:
                    raw = self.rfile.readline(65537)
                    if len(raw) > 65536 or not raw.endswith(b"\n"):
                        return
                    request = json.loads(raw)
                    if not isinstance(request, dict):
                        return
                    response = supervisor.dispatch(request)
                except (OSError, ValueError, KeyError, TypeError) as error:
                    response = {"ok": False, "error": str(error)}
                try:
                    self.wfile.write(json.dumps(response).encode("utf-8") + b"\n")
                except OSError:
                    pass

        server = LocalServer(("127.0.0.1", 0), Handler)
        endpoint = {"protocol": PROTOCOL, "pid": os.getpid(), "port": server.server_address[1],
                    "token": supervisor.token}
        temporary = directory / "endpoint.tmp"
        temporary.write_text(json.dumps(endpoint), encoding="utf-8")
        os.replace(temporary, endpoint_path)
        threading.Thread(target=server.serve_forever, kwargs={"poll_interval": .2}, daemon=True).start()
        server_started = True
        while not supervisor.stopping.wait(.25):
            with supervisor.lock:
                manager.poll()
    finally:
        if server_started:
            server.shutdown()
        if server:
            server.server_close()
        if manager:
            manager.close()
        endpoint_path.unlink(missing_ok=True)
        lock.close()


def main():
    parser = argparse.ArgumentParser(description="Port Forward TUI background supervisor")
    parser.add_argument("--data-dir", type=Path, default=DATA_DIR)
    parser.add_argument("--serve", action="store_true")
    parser.add_argument("--status", action="store_true")
    parser.add_argument("--shutdown", action="store_true")
    options = parser.parse_args()
    if options.serve:
        serve(options.data_dir)
    elif options.shutdown:
        exchange(options.data_dir, "shutdown")
        print("Background manager and its tunnels stopped.")
    else:
        result = exchange(options.data_dir, "status")
        print(json.dumps({k: result[k] for k in ("pid", "host", "running", "states")}, indent=2))


if __name__ == "__main__":
    main()

File: src/background.rs
//! Compatible protocol-1 controller. Views may come and go; this process owns SSH.
use crate::{
    forwarding::Manager,
    process::{Backend, NativeBackend},
    store::{self, Forward, Lock, Store},
};
use anyhow::{Context, Result, ensure};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};
use std::{
    collections::BTreeMap,
    fs::{self, OpenOptions},
    io::Write,
    net::{Ipv4Addr, SocketAddr, TcpListener, TcpStream},
    path::{Path, PathBuf},
    process::{Child, Command, Stdio},
    sync::{
        Arc, Mutex,
        atomic::{AtomicBool, AtomicUsize, Ordering},
    },
    thread,
    time::{Duration, Instant},
};

pub const PROTOCOL: u8 = 1;
const MAX_REQUEST: usize = 65_536;
const MAX_RESPONSE: usize = 2 * 1024 * 1024;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Endpoint {
    pub protocol: u8,
    pub pid: u32,
    pub port: u16,
    pub token: String,
}

fn read_line(stream: &mut TcpStream, limit: usize, timeout: Duration) -> Result<Vec<u8>> {
    use std::io::Read;
    let mut bytes = Vec::new();
    let deadline = Instant::now() + timeout;
    let mut chunk = [0_u8; 8192];
    loop {
        let remaining = deadline.saturating_duration_since(Instant::now());
        ensure!(!remaining.is_zero(), "Controller message timed out.");
        stream.set_read_timeout(Some(remaining))?;
        let count = stream.read(&mut chunk)?;
        ensure!(count != 0, "Incomplete controller message.");
        let end = chunk[..count]
            .iter()
            .position(|byte| *byte == b'\n')
            .map(|offset| offset + 1);
        bytes.extend_from_slice(&chunk[..end.unwrap_or(count)]);
        ensure!(bytes.len() <= limit, "Oversized controller message.");
        if end.is_some() {
            return Ok(bytes);
        }
    }
}

pub fn exchange(
    directory: &Path,
    command: &str,
    arguments: Value,
    timeout: Duration,
) -> Result<Value> {
    let bytes =
        fs::read(directory.join("endpoint.json")).context("No background manager is available.")?;
    ensure!(bytes.len() <= 8192, "Invalid controller endpoint.");
    let endpoint: Endpoint = serde_json::from_slice(&bytes)?;
    ensure!(
        endpoint.protocol == PROTOCOL && endpoint.port > 0 && endpoint.token.len() == 64,
        "Incompatible controller endpoint."
    );
    let address = SocketAddr::from((Ipv4Addr::LOCALHOST, endpoint.port));
    let mut connection = TcpStream::connect_timeout(&address, timeout.min(Duration::from_secs(2)))?;
    connection.set_read_timeout(Some(timeout))?;
    connection.set_write_timeout(Some(timeout))?;
    let mut request = arguments.as_object().cloned().unwrap_or_default();
    request.insert("protocol".into(), json!(PROTOCOL));
    request.insert("token".into(), json!(endpoint.token));
    request.insert("command".into(), json!(command));
    let mut bytes = serde_json::to_vec(&request)?;
    bytes.push(b'\n');
    ensure!(
        bytes.len() <= MAX_REQUEST,
        "Controller request is too large."
    );
    connection.write_all(&bytes)?;
    let response: Value =
        serde_json::from_slice(&read_line(&mut connection, MAX_RESPONSE, timeout)?)?;
    ensure!(
        response.get("ok") == Some(&Value::Bool(true)),
        "{}",
        response
            .get("error")
            .and_then(Value::as_str)
            .unwrap_or("Background manager rejected the request.")
    );
    Ok(response)
}

pub struct Supervisor<B: Backend> {
    pub store: Store,
    pub manager: Manager<B>,
    token: String,
    pub stopping: bool,
}

impl<B: Backend> Supervisor<B> {
    pub fn new(store: Store, backend: B) -> Self {
        Self {
            manager: Manager::new(store.settings.clone(), backend),
            store,
            token: format!("{}{}", store::new_id(), store::new_id()),
            stopping: false,
        }
    }
    pub fn snapshot(&self) -> Value {
        let now = Instant::now();
        let states = self.manager.states();
        let details: BTreeMap<_, _> = states
            .keys()
            .map(|id| (id.clone(), self.manager.details(id, now)))
            .collect();
        json!({"ok":true,"protocol":PROTOCOL,"pid":std::process::id(),
            "host":self.manager.settings.host,"ssh_port":self.manager.settings.ssh_port,"ssh_config":self.manager.settings.ssh_config,
            "capabilities":["shared_favorites","auto_reconnect"],"forwards":self.store.settings.forwards,
            "running":self.manager.running(),"states":states,"details":details})
    }
    pub fn local(&mut self, command: &str, arguments: Value) -> Result<Value> {
        let mut request = arguments;
        request["command"] = json!(command);
        request["protocol"] = json!(PROTOCOL);
        request["token"] = json!(self.token);
        self.dispatch(&request)
    }
    fn authenticate(&self, request: &Value) -> Result<()> {
        let provided = request
            .get("token")
            .and_then(Value::as_str)
            .unwrap_or("")
            .as_bytes();
        ensure!(provided.len() == self.token.len(), "Unauthorized");
        let different = provided
            .iter()
            .zip(self.token.bytes())
            .fold(0_u8, |bits, (left, right)| bits | (left ^ right));
        ensure!(different == 0, "Unauthorized");
        ensure!(
            request.get("protocol") == Some(&json!(PROTOCOL)),
            "Incompatible background protocol; restart the background manager."
        );
        Ok(())
    }
    pub fn dispatch(&mut self, request: &Value) -> Result<Value> {
        self.authenticate(request)?;
        ensure!(
            !self.stopping,
            "Background manager is stopping. Reopen after it exits."
        );
        let action = request
            .get("command")
            .and_then(Value::as_str)
            .context("Missing controller command.")?;
        // Stop works even with damaged JSON or an externally changed destination.
        if !matches!(action, "stop" | "stop_all" | "shutdown") {
            let current = Store::load(&self.store.directory)?;
            self.manager.check_destination(&current.settings)?;
            self.store = current;
        }
        let mut result_rule = None;
        let now = Instant::now();
        match action {
            "status" => {}
            "start" => {
                let id = request
                    .get("rule_id")
                    .and_then(Value::as_str)
                    .context("Missing favorite ID.")?;
                let rule = self
                    .store
                    .settings
                    .forwards
                    .iter()
                    .find(|rule| rule.id == id)
                    .context("Save the forward before starting it.")?;
                self.manager.start(rule, now)?;
            }
            "upsert" => {
                let mut rule: Forward = serde_json::from_value(
                    request.get("rule").context("Missing favorite.")?.clone(),
                )?;
                rule.validate()?;
                ensure!(!rule.name.is_empty(), "A favorite needs a name.");
                let current = self
                    .store
                    .settings
                    .forwards
                    .iter()
                    .find(|saved| saved.id == rule.id)
                    .cloned();
                let expected = request.get("expected").filter(|value| !value.is_null());
                if let Some(expected) = expected {
                    ensure!(
                        current
                            .as_ref()
                            .is_some_and(
                                |saved| serde_json::to_value(saved).ok().as_ref() == Some(expected)
                            ),
                        "This favorite changed in another view. Reopen Edit and try again."
                    );
                } else if let Some(current) = &current {
                    ensure!(
                        current == &rule,
                        "This favorite changed in another view. Refresh and try again."
                    );
                }
                let duplicate = self
                    .store
                    .settings
                    .forwards
                    .iter()
                    .find(|saved| {
                        saved.id != rule.id
                            && saved.local_port == rule.local_port
                            && saved.remote_port == rule.remote_port
                    })
                    .cloned();
                if let Some(duplicate) = duplicate {
                    ensure!(
                        current.is_none(),
                        "That mapping is already saved in another favorite."
                    );
                    rule = duplicate;
                } else {
                    let mut updated = self.store.settings.forwards.clone();
                    if let Some(saved) = updated.iter_mut().find(|saved| saved.id == rule.id) {
                        *saved = rule.clone();
                    } else {
                        updated.push(rule.clone());
                    }
                    self.store.save(updated)?;
                    if current.as_ref().is_some_and(|saved| saved != &rule)
                        && self.manager.state(&rule.id).requested()
                    {
                        self.manager.restart(&rule, now)?;
                    }
                }
                if request.get("start") == Some(&Value::Bool(true)) {
                    self.manager.start(&rule, now)?;
                }
                result_rule = Some(rule.id);
            }
            "delete" => {
                let id = request
                    .get("rule_id")
                    .and_then(Value::as_str)
                    .context("Missing favorite ID.")?;
                if let Some(current) = self
                    .store
                    .settings
                    .forwards
                    .iter()
                    .find(|rule| rule.id == id)
                {
                    ensure!(
                        request.get("expected") == Some(&serde_json::to_value(current)?),
                        "This favorite changed in another view. Select it again before deleting."
                    );
                    let updated = self
                        .store
                        .settings
                        .forwards
                        .iter()
                        .filter(|rule| rule.id != id)
                        .cloned()
                        .collect();
                    self.store.save(updated)?;
                    self.manager.stop(id);
                }
            }
            "stop" => self.manager.stop(
                request
                    .get("rule_id")
                    .and_then(Value::as_str)
                    .context("Invalid favorite ID.")?,
            ),
            "stop_all" | "shutdown" => {
                self.manager.close();
                self.stopping = action == "shutdown";
            }
            _ => anyhow::bail!("Unknown controller command."),
        }
        self.manager.poll(now);
        let mut snapshot = self.snapshot();
        if let Some(id) = result_rule {
            snapshot["rule_id"] = json!(id);
        }
        Ok(snapshot)
    }
}

struct EndpointGuard {
    path: PathBuf,
    token: String,
}
impl Drop for EndpointGuard {
    fn drop(&mut self) {
        if fs::read(&self.path)
            .ok()
            .and_then(|bytes| serde_json::from_slice::<Endpoint>(&bytes).ok())
            .is_some_and(|endpoint| endpoint.token == self.token)
        {
            let _ = fs::remove_file(&self.path);
        }
    }
}

pub fn serve(directory: &Path) -> Result<()> {
    let directory = store::absolute(directory)?;
    let _lock = Lock::acquire(&directory, "daemon.lock", Duration::ZERO)?;
    let store = Store::load(&directory)?;
    store::host(&store.settings.host)?;
    let supervisor = Supervisor::new(store, NativeBackend::new()?);
    let listener = TcpListener::bind((Ipv4Addr::LOCALHOST, 0))?;
    listener.set_nonblocking(true)?;
    let endpoint = Endpoint {
        protocol: PROTOCOL,
        pid: std::process::id(),
        port: listener.local_addr()?.port(),
        token: supervisor.token.clone(),
    };
    let endpoint_path = directory.join("endpoint.json");
    store::write_json(&endpoint_path, &endpoint)?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(&endpoint_path, fs::Permissions::from_mode(0o600))?;
    }
    let _endpoint = EndpointGuard {
        path: endpoint_path,
        token: endpoint.token,
    };
    let supervisor = Arc::new(Mutex::new(supervisor));
    let interrupted = Arc::new(AtomicBool::new(false));
    let flag = Arc::clone(&interrupted);
    ctrlc::set_handler(move || {
        flag.store(true, Ordering::Relaxed);
    })?;
    let clients = Arc::new(AtomicUsize::new(0));
    let mut next_poll = Instant::now();
    loop {
        {
            let mut supervisor = supervisor
                .lock()
                .map_err(|_| anyhow::anyhow!("Controller state lock failed."))?;
            if supervisor.stopping || interrupted.load(Ordering::Relaxed) {
                supervisor.manager.close();
                break;
            }
            if Instant::now() >= next_poll {
                supervisor.manager.poll(Instant::now());
                next_poll = Instant::now()
                    + Duration::from_millis(if cfg!(target_os = "macos") { 750 } else { 250 });
            }
        }
        match listener.accept() {
            Ok((mut stream, _)) => {
                if clients.load(Ordering::Relaxed) >= 16 {
                    continue;
                }
                stream.set_read_timeout(Some(Duration::from_secs(2)))?;
                stream.set_write_timeout(Some(Duration::from_secs(2)))?;
                clients.fetch_add(1, Ordering::Relaxed);
                let clients = Arc::clone(&clients);
                let supervisor = Arc::clone(&supervisor);
                thread::spawn(move || {
                    struct Count(Arc<AtomicUsize>);
                    impl Drop for Count {
                        fn drop(&mut self) {
                            self.0.fetch_sub(1, Ordering::Relaxed);
                        }
                    }
                    let _count = Count(clients);
                    let result = (|| -> Result<Value> {
                        let request: Value = serde_json::from_slice(&read_line(
                            &mut stream,
                            MAX_REQUEST,
                            Duration::from_secs(2),
                        )?)?;
                        let mut supervisor = supervisor
                            .lock()
                            .map_err(|_| anyhow::anyhow!("Controller state lock failed."))?;
                        supervisor.dispatch(&request)
                    })();
                    let response = result
                        .unwrap_or_else(|error| json!({"ok":false,"error":format!("{error:#}")}));
                    if let Ok(mut bytes) = serde_json::to_vec(&response) {
                        if bytes.len() >= MAX_RESPONSE {
                            bytes = br#"{"ok":false,"error":"Controller response exceeded its size limit."}"#.to_vec();
                        }
                        bytes.push(b'\n');
                        let _ = stream.write_all(&bytes);
                    }
                });
            }
            Err(error) if error.kind() == std::io::ErrorKind::WouldBlock => {
                thread::sleep(Duration::from_millis(20))
            }
            Err(error) => return Err(error.into()),
        }
    }
    // Let the shutdown response finish before the process returns. A slow peer
    // cannot hold the process indefinitely; all socket operations have deadlines.
    let deadline = Instant::now() + Duration::from_secs(3);
    while clients.load(Ordering::Relaxed) != 0 && Instant::now() < deadline {
        thread::sleep(Duration::from_millis(10));
    }
    Ok(())
}

pub fn launch(directory: &Path) -> Result<Child> {
    let directory = store::absolute(directory)?;
    fs::create_dir_all(&directory)?;
    let path = directory.join("port_forward_tui.background.log");
    if path.metadata().is_ok_and(|data| data.len() > 256 * 1024) {
        // A concurrent launcher may already have rotated the log. Logging must
        // never replace the daemon lock as the authority on who owns tunnels.
        let _ = fs::rename(
            &path,
            directory.join("port_forward_tui.background.previous.log"),
        );
    }
    let log = OpenOptions::new().create(true).append(true).open(path)?;
    let mut command = Command::new(std::env::current_exe()?);
    command
        .args(["--serve", "--data-dir"])
        .arg(directory)
        .stdin(Stdio::null())
        .stdout(log.try_clone()?)
        .stderr(log);
    crate::process::configure_daemon(&mut command);
    command.spawn().context(
        "Could not detach the background manager. Launch Ports from a normal terminal session.",
    )
}

pub fn ensure_daemon(directory: &Path) -> Result<Value> {
    if let Ok(snapshot) = exchange(directory, "status", json!({}), Duration::from_secs(2)) {
        return Ok(snapshot);
    }
    let mut child = launch(directory)?;
    let deadline = Instant::now() + Duration::from_secs(8);
    while Instant::now() < deadline {
        if let Ok(snapshot) = exchange(directory, "status", json!({}), Duration::from_millis(500)) {
            // Reap our losing launcher if it exited. Never terminate the winner.
            let _ = child.try_wait();
            return Ok(snapshot);
        }
        let _ = child.try_wait();
        thread::sleep(Duration::from_millis(100));
    }
    anyhow::bail!(
        "Background manager did not start. See {}",
        directory.join("port_forward_tui.background.log").display()
    )
}

pub fn call(directory: &Path, command: &str, arguments: Value) -> Result<Value> {
    if !matches!(command, "stop" | "stop_all" | "shutdown")
        || !directory.join("endpoint.json").exists()
    {
        ensure_daemon(directory)?;
    }
    exchange(directory, command, arguments, Duration::from_secs(5))
}

pub fn restart(directory: &Path) -> Result<Vec<String>> {
    let snapshot = match exchange(directory, "status", json!({}), Duration::from_secs(2)) {
        Ok(snapshot) => snapshot,
        Err(_) if !directory.join("endpoint.json").exists() => {
            ensure_daemon(directory)?;
            return Ok(Vec::new());
        }
        Err(error) => return Err(error),
    };
    let wanted = snapshot["forwards"]
        .as_array()
        .context("Invalid controller favorites.")?
        .iter()
        .filter_map(|rule| rule["id"].as_str())
        .filter(|id| {
            matches!(
                snapshot["states"][id].as_str(),
                Some("ON" | "CONNECTING" | "RETRYING")
            )
        })
        .map(str::to_owned)
        .collect::<Vec<_>>();
    exchange(directory, "shutdown", json!({}), Duration::from_secs(5))?;
    // Lock release proves the old controller relinquished ownership. Avoid PID
    // reuse guesses and never force-kill a controller during an upgrade.
    let lock = Lock::acquire(directory, "daemon.lock", Duration::from_secs(8))
        .context("The old manager has not exited; it was not force-killed.")?;
    drop(lock);
    ensure_daemon(directory)?;
    for id in &wanted {
        exchange(
            directory,
            "start",
            json!({"rule_id":id}),
            Duration::from_secs(5),
        )?;
    }
    Ok(wanted)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        process::{Listeners, TunnelProcess},
        store::Settings,
    };
    struct NeverSsh;
    impl Backend for NeverSsh {
        fn start(&mut self, _: &Settings, _: &Forward) -> Result<Box<dyn TunnelProcess>> {
            anyhow::bail!("Isolated controller test")
        }
        fn listeners(&mut self, _: &[u32]) -> Result<Listeners> {
            Ok(Default::default())
        }
    }
    fn request(supervisor: &Supervisor<NeverSsh>, command: &str, arguments: Value) -> Value {
        let mut request = arguments;
        request["token"] = json!(supervisor.token);
        request["protocol"] = json!(PROTOCOL);
        request["command"] = json!(command);
        request
    }
    #[test]
    fn stale_concurrent_edits_and_deletes_cannot_overwrite_newer_favorites() {
        let temp = tempfile::tempdir().unwrap();
        let mut store = Store::load(temp.path()).unwrap();
        store.settings.host = "workbox".into();
        store.save(vec![]).unwrap();
        let mut supervisor = Supervisor::new(store, NeverSsh);
        let first = Forward::new(18000, 8000, "API").unwrap();
        let req = request(&supervisor, "upsert", json!({"rule":first,"expected":null}));
        supervisor.dispatch(&req).unwrap();
        let mut edited = first.clone();
        edited.name = "Renamed API".into();
        let req = request(
            &supervisor,
            "upsert",
            json!({"rule":edited,"expected":first}),
        );
        supervisor.dispatch(&req).unwrap();
        let req = request(
            &supervisor,
            "upsert",
            json!({"rule":first,"expected":first}),
        );
        assert!(supervisor.dispatch(&req).is_err());
        let req = request(
            &supervisor,
            "delete",
            json!({"rule_id":first.id,"expected":first}),
        );
        assert!(supervisor.dispatch(&req).is_err());
        assert_eq!(
            Store::load(temp.path()).unwrap().settings.forwards,
            [edited]
        );
        let duplicate = Forward::new(18000, 8000, "Second name").unwrap();
        let req = request(&supervisor, "upsert", json!({"rule":duplicate}));
        let snapshot = supervisor.dispatch(&req).unwrap();
        assert_eq!(snapshot["rule_id"], first.id);
        assert_eq!(snapshot["forwards"].as_array().unwrap().len(), 1);
    }
    #[test]
    fn authentication_fails_before_mutation_and_stop_survives_broken_json() {
        let temp = tempfile::tempdir().unwrap();
        let mut store = Store::load(temp.path()).unwrap();
        store.settings.host = "workbox".into();
        store.save(vec![]).unwrap();
        let mut supervisor = Supervisor::new(store, NeverSsh);
        assert!(
            supervisor
                .dispatch(&json!({"protocol":1,"token":"bad","command":"shutdown"}))
                .is_err()
        );
        assert!(!supervisor.stopping);
        fs::write(temp.path().join("forwards.json"), b"broken").unwrap();
        let req = request(&supervisor, "status", json!({}));
        assert!(supervisor.dispatch(&req).is_err());
        let req = request(&supervisor, "stop_all", json!({}));
        assert!(supervisor.dispatch(&req).is_ok());
        let req = request(&supervisor, "shutdown", json!({}));
        assert!(supervisor.dispatch(&req).is_ok());
        assert!(supervisor.stopping);
    }
}
